pilot/adapter/config/file/monitor_test.go

```diff
@@ -0,0 +1,259 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package file_test
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"testing"
+
+	"istio.io/istio/pilot/adapter/config/file"
+	"istio.io/istio/pilot/adapter/config/memory"
+	"istio.io/istio/pilot/model"
+)
+
+var (
+	files = []string{
+		"cb-policy.yaml",
+		"timeout-route-rule.yaml",
+		"weighted-route.yaml",
+		"fault-route.yaml",
+		"redirect-route.yaml",
+		"rewrite-route.yaml",
+		"websocket-route.yaml",
+		"egress-rule.yaml",
+		"egress-rule-cb-policy.yaml",
+		"egress-rule-timeout-route-rule.yaml",
+		"ingress-route-world.yaml",
+		"ingress-route-foo.yaml",
+	}
+)
+
+func createTempDir() string {
+	// Make the temporary directory
+	dir, _ := ioutil.TempDir("/tmp/", "monitor")
+	_ = os.MkdirAll(dir, os.ModeDir|os.ModePerm)
+	return dir
+}
+
+type event struct {
+	config model.Config
+	event  model.Event
+}
+
+type controllerManager struct {
+	controller     model.ConfigStoreCache
+	controllerStop chan struct{}
+}
+
+func (cm *controllerManager) setup(eventCh chan event) {
+	cm.controllerStop = make(chan struct{})
+	store := memory.Make(model.IstioConfigTypes)
+	cm.controller = memory.NewControllerWithBufferSize(store, 100)
+
+	// Register changes to the repository
+	for _, s := range model.IstioConfigTypes.Types() {
+		cm.controller.RegisterEventHandler(s, func(config model.Config, ev model.Event) {
+			eventCh <- event{
+				config: config,
+				event:  ev,
+			}
+		})
+	}
+
+	// Run the controller.
+	go cm.controller.Run(cm.controllerStop)
+}
+
+func (cm *controllerManager) teardown() {
+	close(cm.controllerStop)
+}
+
+type env struct {
+	controllerManager controllerManager
+	monitorCtx        context.Context
+	monitorCancel     context.CancelFunc
+	eventCh           chan event
+	fsroot            string
+	monitor           *file.Monitor
+}
+
+func (e *env) setup(tempDir string) {
+	e.eventCh = make(chan event)
+
+	// Create and setup the controller.
+	e.controllerManager = controllerManager{}
+	e.controllerManager.setup(e.eventCh)
+
+	e.monitorCtx, e.monitorCancel = context.WithCancel(context.Background())
+	e.fsroot = tempDir
+
+	// Make all of the updates go through the controller so that the events are triggered.
+	e.monitor = file.NewMonitor(e.controllerManager.controller, e.fsroot, nil)
+	e.monitor.Start(e.monitorCtx)
+}
+
+func (e *env) teardown() {
+	e.monitorCancel()
+	e.controllerManager.teardown()
+
+	// Remove the temp dir.
+	os.RemoveAll(e.fsroot)
+}
+
+func (e *env) watch() event {
+	return <-e.eventCh
+}
+
+func (e *env) watchFor(eventType model.Event) event {
+	for {
+		if ev := e.watch(); ev.event == eventType {
+			return ev
+		}
+	}
+}
+
+func copyTestFile(fileName string, toDir string) error {
+	return copyFile("testdata/"+fileName, toDir+"/"+fileName)
+}
+
+func copyFile(src, dst string) error {
+	data, err := ioutil.ReadFile(src)
+	if err != nil {
+		return err
+	}
+
```

Comment:
Here and elsewhere delete the newline before the return. No need in for it and it just adds vertical spacing for no real reason. (It overemphasizing the error handling code, which is almost never the important part.)